docs/proposals/server-side-apply.md

```diff
@@ -0,0 +1,237 @@
+
+- More reliable dry-runs (as admission controller is executed) ([ISSUE-804][5])
+- [Syncs always run][2] mutating webhooks (even without diff)
+- [Fix big CRD][3] sync issues
+- Better interoperability with different controllers
+
+Kubernetes SSA Proposal ([KEP-555][13]) has more details about how it works.
+
+## Motivation
+
+ArgoCD uses kubectl library while syncing resources in the cluster. Kubectl uses
+by default a 3-way-merge logic between the live state (in k8s), desired state
+(in git) and the previous state (`last-applied-configuration` annotation) to
+calculate diffs and patch resources in the cluster. This logic is executed in
+the client (ArgoCD) and once the patch is calculated it is then sent to the
+server.
+
+This strategy works well in the majority of the use cases. However, there are
+some scenarios where calculating patches in the client side can cause problems.
+
+Some of the known problems about using client-side approach:
+
+### Better interoperability with Admission Controllers 
+
+More and more users are adopting and configuring [Admission Controllers][4] in
+Kubernetes with different flavors of Validating Webhooks and Mutating Webhooks.
+Admission Controllers will only execute in server-side. In cases where users
+rely on dry-run executions to decide if they should proceed with a deployment,
+having the patch calculated at the client side might provide undesired results.
+Having SSA enabled syncs also guaranties that Admission Controllers are always
+executed, even when there is no diff in the resource.
+
+### Better resource conflict management
+
+Server-Side Apply will better handle and identify conflicts during syncs by
+analyzing the `managedFields` metadata available in all Kubernetes resources
+(since kubernetes 1.18). 
+
+### Better CRD support
+
+By not having to rely on the `last-applied-configuration` annotation, SSA would
+help with failing syncs caused by exceeded annotation size limit. This is a
+common issue when syncing CRDs with large schemas.
+
+## Goals
+
+All following goals should be achieve in order to conclude this proposal:
+
+#### [G-1] Fine grained configuration
+
+- Provide the ability for users to define if they want to use SSA during syncs
+- Users should be able to enable SSA at the controller level (via binary flag)
+(see [UC-1](#uc-1-as-a-user-i-would-like-enable-ssa-at-the-controller-level-so-all-application-are-applied-server-side))
+- Users should be able to enable SSA for a given Application (via syncOptions)
+(see [UC-2](#uc-2-as-a-user-i-would-like-enable-ssa-at-the-application-level-so-all-resources-are-applied-server-side))
+- Users should be able to enable SSA at resource level (via annotation) (see
+[UC-3](#uc-3-as-a-user-i-would-like-enable-ssa-at-the-resource-level-so-only-a-single-manifest-is-applied-server-side)
+- Relates to [ISSUE-2267][6]
+
+#### [G-2] Strategic merge patch while diffing
+
+- Diffing needs to support strategic merge patch (see [ISSUE-2268][7])
```

Comment:
Hello, I think we are hitting this issue https://github.com/kubernetes/kubernetes/issues/39188 when doing diff from ArgoCD. I checked the gitops engine library code and I am wondering what changes should we make on it to support this feature? Thanks! 
 If I understand correctly we need to modify the three way patch code here https://github.com/argoproj/gitops-engine/blob/master/pkg/diff/diff.go#L230?

Reply:
Hi @yeya24. Tks for this!
It would be awesome if you could provide the minimal configuration to reproduce the issue you are having. If you provide that, I'll make sure to update the proposal linking this goal with your use-case.

Reply:
Here is a case borrowed from our user's accident record.

Initial deployed manifests
```yaml
apiVersion: v1
kind: Service
metadata:
    name: multiple-protocol-port-svc
spec:
    ports:
        - name: rtmpk
          port: 1986
          protocol: UDP
          targetPort: 1986
        - name: rtmp
          port: 1935
          targetPort: 1935
        - name: https
          port: 443
          targetPort: 443
```

updated and expecting manifest

```yaml
apiVersion: v1
kind: Service
metadata:
    name: multiple-protocol-port-svc
spec:
    ports:
        - name: rtmpk
          port: 1986
          protocol: UDP
          targetPort: 1986
        - name: rtmp
          port: 1935
          targetPort: 1935
        - name: rtmpq
          port: 1935
          protocol: UDP
          targetPort: 1935
        - name: https
          port: 443
          targetPort: 443
        - name: http3
          port: 443
          protocol: UDP
          targetPort: 443
```
The actually sync results in live manifest

```yaml
apiVersion: v1
kind: Service
metadata:
    name: multiple-protocol-port-svc
spec:
    ports:
        - name: rtmpk
          port: 1986
          protocol: UDP
          targetPort: 1986
        - name: rtmp
          port: 1935
          targetPort: 1935
          protocol: UDP
        - name: https
          port: 443
          targetPort: 443
          protocol: UDP
```
This mis-behavior not just makes the update results wrong. But also disabled the in-using 443-tcp, which caused very serious prod accidents.
This bug definitely would hurt the usability of the argocd as the quic protocol becomes more and more important. Now more and more users are updating they software to adopt the quic, which need to enable the udp protocol on the same port with tcp. Then they will be blocked by this bug, and possibly results in serious accident.

Reply:
> Hi @yeya24. Tks for this! It would be awesome if you could provide the minimal configuration to reproduce the issue you are having. If you provide that, I'll make sure to update the proposal linking this goal with your use-case. This way we can use it as a test case by the time we are implementing the feature. 👍🏻

@leoluz  Here is a case borrowed from our user's accident record.

Initial deployed manifests
```yaml
apiVersion: v1
kind: Service
metadata:
    name: multiple-protocol-port-svc
spec:
    ports:
        - name: rtmpk
          port: 1986
          protocol: UDP
          targetPort: 1986
        - name: rtmp
          port: 1935
          targetPort: 1935
        - name: https
          port: 443
          targetPort: 443
```

updated and expecting manifest

```yaml
apiVersion: v1
kind: Service
metadata:
    name: multiple-protocol-port-svc
spec:
    ports:
        - name: rtmpk
          port: 1986
          protocol: UDP
          targetPort: 1986
        - name: rtmp
          port: 1935
          targetPort: 1935
        - name: rtmpq
          port: 1935
          protocol: UDP
          targetPort: 1935
        - name: https
          port: 443
          targetPort: 443
        - name: http3
          port: 443
          protocol: UDP
          targetPort: 443
```
The actually sync results in live manifest

```yaml
apiVersion: v1
kind: Service
metadata:
    name: multiple-protocol-port-svc
spec:
    ports:
        - name: rtmpk
          port: 1986
          protocol: UDP
          targetPort: 1986
        - name: rtmp
          port: 1935
          targetPort: 1935
          protocol: UDP
        - name: https
          port: 443
          targetPort: 443
          protocol: UDP
```
This mis-behavior not just makes the update results wrong. But also disabled the in-using 443-tcp, which caused very serious prod accidents.
This bug definitely would hurt the usability of the argocd as the quic protocol becomes more and more important. Now more and more users are updating they software to adopt the quic, which need to enable the udp protocol on the same port with tcp. Then they will be blocked by this bug, and possibly results in serious accident.